source/Calamari.Common/Features/Processes/Semaphores/SystemSemaphoreManager.cs

```diff
@@ -90,24 +95,32 @@ IDisposable AcquireMutex(string name, string waitMessage)
             }
 
             return new Releaser(() =>
-            {
-                mutex.ReleaseMutex();
-                mutex.Dispose();
-            });
+                                {
+                                    mutex.ReleaseMutex();
+                                    mutex.Dispose();
+                                });
         }
 
-        static Semaphore CreateGlobalSemaphoreAccessibleToEveryone(string name)
+        Semaphore CreateGlobalSemaphoreAccessibleToEveryone(string name)
         {
             var semaphoreSecurity = new SemaphoreSecurity();
             var everyone = new SecurityIdentifier(WellKnownSidType.WorldSid, null);
             var rule = new SemaphoreAccessRule(everyone, SemaphoreRights.FullControl, AccessControlType.Allow);
 
             semaphoreSecurity.AddAccessRule(rule);
 
-            bool createdNew;
+            var semaphore = new Semaphore(1, 1, name);
+
+            try
+            {
+                semaphore.SetAccessControl(semaphoreSecurity);
+            }
+            catch (Exception e)
+            {
+                log.Verbose($"Failed to set access controls on semaphore '{name}': {e.PrettyPrint()}");
+                //at this point we have a valid semaphore, which is what we'd end up doing anyway
+            }
```

Comment:
If we had created the semaphore on line 112 and this failed, then exception handler would have attempted to create/acquire the exact same semaphore again, just without this access control call.
I figured if this SetAccessControl failed, then we still have a valid semaphore, or at least the exact same semaphore we'd get in the exception handler

Reply:
I wonder if there's any side effect from `SetAccessControl` failing...

Otherwise you might as well do this operation in 2 stages:
1. Create the Semaphore in `AcquireSemaphore()` in a retry loop (does this ever fail?)
2. Try to set the access control in a retry loop

Reply:
> Otherwise you might as well do this operation in 2 stages:

Yeah, good point.

> Create the Semaphore in AcquireSemaphore() in a retry loop (does this ever fail?)

This is the bit that is throwing the exception, not the `SetAccessControl`

Reply:
@flin-8 I looked into this, `SetAccessControl` calls through to underlying Win32 APIs to set ACL's on the semaphore

It uses the `NativeObjectSecurity.Persist` method.

I think, considering that all the errors we see are around creating/acquiring the semaphore, that we can probably do the create/acquire in a retry loop, then just handle the access control separately

Reply:
Changed in [045e62b](https://github.com/OctopusDeploy/Calamari/pull/1414/commits/045e62bcbfee1af5fa152a58fb517869f1c7484f)